WordPress/gutenberg
# REST API: new endpoint to fetch post counts by post status

## What?

Creates a new endpoint to fetch the total number of post counts by post status using [wp_count_posts](https://developer.wordpress.org/reference/functions/wp_count_posts/).


## Why?

The need for a performant post counts check for the block editor. See: https://github.com/WordPress/gutenberg/pull/65223#issuecomment-2381323815

Contrast this with `$query->found_posts` which returns the total count for a specific query only and would degrade performance given `n` posts.

## How?

Following the simplicity of [wp/v2/statuses](https://github.com/wordpress/wordpress-develop/blob/309ecbd3241895522cd6d1eb0ef7f176dfceea94/src/wp-includes/rest-api/endpoints/class-wp-rest-post-statuses-controller.php), which wraps [get_post_stati](https://developer.wordpress.org/reference/functions/get_post_stati/), creating a basic route to return post type counts, e.g., `wp/v2/counts/<post_type>`


## Testing Instructions

Run the tests: `npm run test:unit:php:base -- --filter Gutenberg_Test_REST_Post_Counts_Controller`

Open up the site editor and have a play with the console:

Examples:

```js

await wp.apiFetch( { path: '/wp/v2/counts/page' } )

/*
{
    "publish": 1,
    "future": 0,
    "draft": 2,
    "pending": 0,
    "private": 0,
    "trash": 0,
    "auto-draft": 0
}
*/

await wp.apiFetch( { path: '/wp/v2/counts/wp_block' } )

/*
{
    "publish": 2,
    "future": 0,
    "draft": 0,
    "pending": 0,
    "private": 0,
    "trash": 0,
    "auto-draft": 0
}
*/
```

Also add a custom post type and ensure it appears in the response:

```php
function add_my_custom_post_type() {
	register_post_status(
		'sweaty-palms',
		array(
			'label'                     => _x( 'Sweaty palms', 'post' ),
			'label_count'               => _n_noop(
				'Sweaty palms <span class="count">(%s)</span>',
				'Sweaty palms <span 
class="count">(%s)</span>'
			),
			'public'                    => true,
			'exclude_from_search'       => false,
			'show_in_admin_all_list'    => true,
			'show_in_admin_status_list' => true,
		)
	);
}
add_action( 'init', add_my_custom_post_type );

```
